ovlfs-2.0.1/doc/ovl.8
.\" COPYRIGHT (C) 1998-2003 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: ovl.m
.\"
.\" FILE VERSION:
.\"	@(#)ovl.m	1.4	[03/07/27 22:20:33]
.\"
.\" DESCRIPTION: Manual page for the overlay filesystem.
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 03/08/1998	ARTHUR N.	Initial Release
.\" 03/10/1998	ARTHUR N.	Changed tmac.an to tmac.asn to remove the
.\"				 conflict with the man page tmac.an file.
.\" 03/10/1998	ARTHUR N.	Added the shortcuts for options.
.\" 06/01/2003	ARTHUR N.	Updated for the new release, and made other,
.\"				 minor enhancements.
.\"
.\"
.\" M4 definitions:
.\" COPYRIGHT (C) 1998 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: tmac.asn
.\"
.\" DESCRIPTION: This file contains general purpose nroff/troff macros.
.\"
.\" SCCS ID: @(#) tmac.asn 1.1
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 03/08/1998	ARTHUR N.	Initial Release
.\"
.\"
.\" **************************************************************************
.\"
.\" START OF LIST MACROS
.\"
.de BL    \" Begin Bullet List
.in +8n             \" increase indent by about 8 characters
.nr *i \\n(.i/1n    \" remember the current indent in units of 'en'
.nr *i +2           \" add two more characters to the list item indent
.ds OL \\*(LM       \" remember the last mark - can only nest 2 levels
.ds LM \(bu         \" set the list item marker to a bullet
..
.de LE    \" End List
.nr *i -10      \" calculate original indent
.in \\n(*in     \" reset original indent
.ds LM \\*(OL   \" restore last marker saved
..
.de LI    \" List Item
.in \\n(*in    \" indent to list item space
.br            \" make sure to start a new line
.ti -2n        \" indent, next line only, back about 2 characters
\\*(LM         \" display list marker
..
.\"
.\" **************************************************************************
.\"
.\" START OF THE HANGING INDENT MACRO
.\"
.\" begin hanging indent from current point after the text of argument 1,
.\"  if defined.
.\"
.nr %I 0
.nr w 1n	\" Set this once and allow it to be changed later
.de HI
.if \\n(%I-9  .ab HI macro stack error    \" If %I > 9, can't continue
.nr I\\n(%I \\n(.iu
.nr %I \\n(%I+1
.di *H           \" Divert output to *H macro and put $1 on input
\&\\$1
.br              \" Break this input to put in output (so the diversion gets it)
.di              \" End the diversion
.nr $i \\n(dlu+\\nwu  \" Calculate the hanging indentation
.nr %i \\nw-\w'  '    \" Take offset and subtract the spaces that are coming
.ti 0
\&\\*(*H\\h'\\n(%iu'
'in \\n($iu
..
.\"
.\" end of hanging indent - restore the PREVIOUS indent
.\"
.de HE
.nr %I \\n(%I-1       \" Decrement current indentation counter
.in \\n(I\\n(%Iu      \" Restore indentation at current counter
..
.\"
.\" END OF THE HANGING INDENT MACRO
.\"
.\" **************************************************************************
.\"
.\" START OF THE TAB FIELD MACRO
.\"
.\" The .Tf macro moves over to the specified horrizontal position on the line
.\"  If the current input is past the requested position, then a break is
.\"  inserted followed by the movement to the requested position.
.\"
.\" NOTE: the positioning does not work correctly for troff; I believe the
.\"       problem is the length of the diversion is different than the length
.\"       of the string when output.  The results are close, but not perfectly
.\"       aligned
.\"
.de Tf  \" tab field
.\"		Check argument
.if \\n(.$<1 .ab Tf macro requires an argument
.\"		Find the length of the current line: use a diversion to get it
.di }s
.br
.di
.\"		Find length of the diversion plus the trailing space that will
.\"		follow it
.di }t
'ti 0           \"	Indent already part of string }s
\\*(}s\c\ 
.br
.di
.\"		Calculate the length of the line (minus the current indent)
.nr }u \\n(dlu-\\n(.iu
.br
.\"		Now put the original input line back out followed by the
.\"		horizontal positioning requested
'ti 0
\\*(}s\c            \"		Again, indent is already part of string }s
.ie \\n(}u-\\$1>0 \{
.br
\\h'\\$1-1m'\c \}
.el \\h'\\$1-\\n(}uu-1m'\c
..
.\"
.\"
.\"
.ad l
.TH ovl 8 "Sun Jul 27 22:21:37 EDT 2003" "Arthur Naseef" "Linux System Admin"
.\"
.\" SECTION BREAK
.\"
.SH NAME
ovl.o -
overlay pseudo filesystem.
.\"
.\" SECTION BREAK
.\"
.SH VERSION
This manual page covers version 2.0.1 of the overlay filesystem.
.\"
.\" SECTION BREAK
.\"
.SH SYNOPSIS
modprobe ovl.o
.br
mount -t ovl \fIovl0\fR /mnt -o root=\fIrootdir1\fR,storage=\fIrootdir2\fR
.\"
.\" SECTION BREAK
.\"
.SH DESCRIPTION
The overlay filesystem is a pseudo filesystem that allows any filesystem,
including one that is read-only, to be modified by saving the changes in
another filesystem.
.P
There is much documentation on the issues of using the overlay filesystem
that is included with the source distribution.
This manual page is intended only to clarify how to use the filesystem.
.\"
.\" SECTION BREAK
.\"
.SH USAGE
In order to use the overlay filesystem,
determine the root directories of the base and storage filesystems,
and any special mount options to use.
Then, the mount command may be used to mount the filesystem.
As shown in the SYNOPSIS above,
use the mount command to mount the filesystem.
The device name is not important, and can be chosen at your leisure,
because the filesystem does not use a device.
.P
For example, if the root of the base filesystem is \fB/cdrom\fR
the root of the storage filesystem is \fB/cdrom_ovl\fR,
the storage file is \fB/.cdrom_stg\fR,
and the mount point is \fB/mnt\fR,
then the following mount command can be used:
.P
.in +4n
.hy 0
.HI /sbin/mount
\-t ovl cdrom_ovl \fB/mnt\fR \-o root=\fB/cdrom\fR,storage=\fB/cdrom_ovl\fR,\
stg_file=\fB/.cdrom_stg\fR
.HE
.hy 1
.in
.\"
.\" SUB-SECTION BREAK
.\"
.SS MOUNT OPTIONS
Many mount options are available
with the ovlfs.
Here are the general-purpose
mount options:
.P
.in +1i
base_root,
storage_root,
stg_method,
nostorage,
maxmem,
xmnt,
noxmnt,
updmntonly,
and noupdmntonly.
.in
.P
Options which affect
the storage of inode mappings,
if compiled
into the driver,
are listed below:
.P
.in +1i
storemaps,
nostoremaps,
basemap,
nobasemap,
stgmap,
and nostgmap
.in
.P
Last, the following options are available for the magic directories,
if compiled into the driver:
.P
.in +1i
magic,
nomagic,
hidemagic,
showmagic,
basemagic,
nobasemagic,
ovlmagic,
noovlmagic,
smagic,
and bmagic
.in
.P
Note that there are shortcuts for these options;
see the subsection, \fBShortcuts\fR, for more information.
Each option is described
in more detail here:
.TP
root=\fIdir\fR
Set the root of the base filesystem to the named directory;
this is the only way to specify this information.
The string \fBbase_root=\fR may be used in place of root=.
.TP
storage=\fIdir\fR
Set the root of the storage filesystem to the named directory;
this is the only way to specify this information.
The string \fBstorage_root=\fR may be used in place of storage=.
.TP
stg_file=\fIpath\fR
Set the path to the storage file to the named path.
All inode information and inode mapping information is
saved between mounts in the storage file.
.TP
nostorage
Turn off the use of storage.
This option does not work yet.
.TP
maxmem
Turn on checking of the amount of available memory before
allocating more memory for inode storage for the overlay filesystem.
From testing, it seems that this option is better left off.
.TP
xmnt
Turn on the crossing of mount points in the base
and storage filesystems.
See the file, Options, distributed with
the source tree for more information.
.TP
noxmnt
Turn off the crossing of mount points in the base
and storage filesystems.
See the file, Options, distributed with
the source tree for more information.
.TP
storemaps
Store all inode mapping information in the storage file.
.TP
nostoremaps
Do not store any inode mapping information in the storage file.
.TP
basemap
Store the inode mapping information for the base filesystem
in the storage file.
.TP
nobasemap
do not store the inode mapping information for the base filesystem
in the storage file.
.TP
stgmap
Store the inode mapping information for the storage filesystem
in the storage file.
.TP
nostgmap
Do not store the inode mapping information for the storage filesystem
in the storage file.
.TP
hidemagic
Do not display the magic directories in directory listings.
In other words, with this option,
the readdir() function will not include the magic directories.
.TP
showmagic
Display the magic directories in directory listings.
This is the opposite of the hidemagic option.
.TP
magic
Turn on the default magic directory options.
At the time of this writing, the default options include
basemagic, ovlmagic, and hidemagic.
.TP
basemagic
Turn on the magic directory for the base filesystem.
.TP
nobasemagic
Turn off the magic directory for the base filesystem.
.TP
ovlmagic
Turn on the magic directory for the storage filesystem.
.TP
noovlmagic
Turn off the magic directory for the storage filesystem.
.TP
smagic=\fIdir\fR
Set the name of the magic directory for the storage filesystem
to the specified name.
Do not include any slashes in this name.
The option, \fBmagic=\fIdir\fR, is a synonym for this option.
.TP
bmagic=\fIdir\fR
Set the name of the magic directory for the base filesystem
to the specified name.
Do not include any slashes in this name.
The option, \fBmagicr=\fIdir\fR, is a synonym for this option.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Shortcuts
All of the supported mount options have shortcuts to
help reduce the length of the mount options on the command line.
.P
.de li
.LI
\\$1
.Tf 24n
\\$2
..
.BL
.li nostorage nost
.li maxmem mm
.li xmnt xm
.li noxmnt noxm
.li updmntonly um
.li noupdmntonly noum
.li base_root=<dir> br=<dir>
.li storage_root=<dir> sr=<dir>
.li stg_file=<path> sf=<path>
.li stg_method=<path> method=<path>
.li storemaps ma
.li nostoremaps noma
.li basemap bm
.li nobasemap nobm
.li stgmap sm
.li nostgmap nosm
.li hidemagic hm
.li magic mg
.li nomagic nomg
.li basemagic mb
.li nobasemagic nomb
.li ovlmagic mo
.li noovlmagic nomo
.li showmagic ms
.li smagic=<name> sn=<name>
.li bmagic=<name> bn=<name>
.LE
.\"
.\" SECTION BREAK
.\"
.SH WARNINGS
The overlay filesystem in many ways emulates the VFS.
There are many possible ways that the use of the overlay
filesystem may cause problems.
A few of the most important issues are mentioned here.
Please see the rest of the documentation included with the source
for more information.
.P
Note that in spite of all of these warnings, and all others
documented for the overlay filesystem,
the use of the overlay filesystem should not cause any corruption
to any of the filesystems involved.
However, for your own safety, it is a good idea to backup ALL
filesystems on the host before using the overlay filesystem.
.P
.BL
.LI
Do not attempt to use a storage root that is a subdirectory of the
base root directory, or vice versa.
.P
.LI
All inodes that are not regular files or directories,
such as device files, pipes, and sockets,
may be directly accessed for reading and writing
through the overlay filesystem.
.P
.LI
The access and change times of the inodes in the base filesystem
may be affected by the use of the overlay filesystem,
although the modification times should not be affected -
except for files that are not directories or regular files.
.P
.LI
Any change to a file in the base filesystem causes the entire
file to be copied into the storage filesystem, even if only
one byte of the file is modified, and regardless of the size
of the file.
.P
.LI
Since names are only pointers to inodes, files and directories that
need to be created in the storage filesystem to store changes from
the base filesystem, may take a name other than the one being
accessed in the base filesystem.
This should only happen with files that have more than one hard link.
.P
.LI
Removing files from the overlay filesystem does not remove their inodes.
The only way to cleanup old inodes in the overlay filesystem is to remove
the storage file, or use a new storage file, for the filesystem.
.LE
.\"
.\" SECTION BREAK
.\"
.SH SEE ALSO
mount(8), ovlfsmount(8), ovlfstab(5)
.\"
.\" SECTION BREAK
.\"
.SH COPYRIGHT
This manual page and associated source code is licensed under the
GNU General Public License Version 2.
Of course, all source derived from the linux kernel source is copyrighted
by the original author of that source.
.sp 1
COPYRIGHT (C) 1998-2003 By Arthur Naseef


ovlfs-2.0.1/RELEASE_NOTES
                               ==================
                               OVERLAY FILESYSTEM
                                 RELEASE NOTES
                                 VERSION 2.0.0
                               ==================


1. ROBUSTNESS
=============

    Due to the amount of time passed since the initially planned release of
the 2.0.0 version, the following has been affected:

	A. Some of the documentation is not up-to-date.

	B. New files and code have been added with no real impact on the
	   software.

	C. Testing with the latest version of the 2.4 kernel has not been
	   completed.

	D. New features are not documented.

	E. Old warnings and issues may not be applicable.

    In spite of these, this software is working and stable enough for release.
I hope you will find this to be the case.

    Note that support for the ReiserFS as the base or storage filesystem is
not complete; multiple hard links to files, and renamed files in the base or
storage filesystem may not be properly resolved by the Overlay Filesystem.
Otherwise, ReiserFS support is functional.



2. KERNEL VERSIONS
==================

    This version of the Overlay Filesystem is built on, and supports, 2.4
series kernels.  Please note that a limited set of 2.4 kernels have actually
been used, so there may be incompatibilities with others due to changes in the
VFS or other portions of the Linux kernel.

    While support for 2.2 kernels was considered, it is not complete; the 2.2
and 2.4 kernels are very similar, but 2.2 lacks many useful interfaces on which
the Overlay Filesystem relies.  Therefore, 2.2 support is currently incomplete.

    Support for 2.0 kernels was present in the 1.0 series for the Overlay
Filesystem, but that support has been lost in the migration to 2.4; in
addition, the 1.0 ovlfs versions had serious flaws that have been corrected.



3. BUG FIXES
============

    Since 2.0.0 is a major rewrite from version 1, listing bug fixes is not
effective.  However, there were serious flaws in version 1 that have been
corrected.  These are listed here:

	A. Directory hierarchy creation used a recursive method that could
	   easily overflow the kernel stack.

	B. Large stack variables were allocated by the ovlfs, which could
	   easily overflow the kernel stack.  All such variables have been
	   replaced with heap memory.

	C. A storage filesystem root that could be reached under the base
	   filesystem root was not supported, and such an instance lead to
	   certain, serious problems in the kernel.  This is no longer true.



4. IMPROVEMENTS
===============

    Many improvements have been made to the Overlay Filesystem.  The majority
support future development efforts and improve the stability and operation of
the filesystem.

    The most notable improvement is the Storage System support which allows
alternative methods of storing Overlay Filesystem data to be added to the core
without changing the core.  Support for storing the filesystem's data on block
devices and in a block-formatted file are planned, which will have a great
impact on the filesystem's required memory use.

    Another notable improvement is the ovlmount program included in this
release.  It simplifies the management of multiple Overlay Filesystem mounts
by using an fstab-like file specifically designed for the Overlay Filesystem.



5. SMP SUPPORT
==============

    At this time, SMP support is not tested, and is expected NOT to work.
The effort needed to find and resolve SMP issues has not been made yet.



6. DEVELOPER'S NOTE
===================

    Much remains to be cleaned up for "correct" kernel coding.  This holds
especially true in the case of module versions.  The current plan for multiple
kernel version support involves splitting the core into different directories,
with one for each major implementation based on the kernel version.  With this
model, the existing code can be greatly reduced in size and readability.

    Any help in this matter is greatly appreciated.



7. LIVE CD CREATED
==================

    With this version of the Overlay Filesystem, I have successfully created a
"live", bootable CD using the Overlay Filesystem as the root filesystem.  Note
that the real root is replaced after kernel startup using a small application
that I wrote.

    The CD boots directly to a SuSE 7.3 system with web browser and all types
of wonderful features.  If you would like to know more, please write!

    One of the great benefits of the live CD is the thorough testing it gives
the Overlay Filesystem!



8. LEX AND YACC
===============

    The mount directory uses lex and yacc (or flex and bison) to generate a
parser for the ovlfstab file.  Output from these programs is included in the
release by default, and will not be removed by "make clean", so that lex and
yacc should not be required on your system.  Note, though, that the flex
library is needed.

    If you prefer to rebuild the lex and yacc output, please note that the
makefiles expect programs named lex and yacc by default.



9. THANKS
=========

    Many thanks goes out to the User Mode Linux project hosted on Sourceforge
at the following URL:

	http://user-mode-linux.sourceforge.net/

Debugging the Overlay Filesystem has been made significantly easier, and
corruption resulting from bugs was limited to a "safe area" (i.e. the content
of a single file).

    Keep up the good work!



[version @(#) RELEASE_NOTES 1.3]


ovlfs-2.0.1/doc/ovl.m
.\" COPYRIGHT (C) 1998-2003 By Arthur Naseef
.\"
.\" This file is covered under the GNU General Public License Version 2.  For
.\"  more information, see the file COPYING.
.\"
.\"
.\" FILE: ovl.m
.\"
.\" FILE VERSION:
.\"	@(#)ovl.m	1.4	[03/07/27 22:20:33]
.\"
.\" DESCRIPTION: Manual page for the overlay filesystem.
.\"
.\"
.\" REVISION HISTORY
.\"
.\" DATE	AUTHOR		DESCRIPTION
.\" ===========	===============	==============================================
.\" 03/08/1998	ARTHUR N.	Initial Release
.\" 03/10/1998	ARTHUR N.	Changed tmac.an to tmac.asn to remove the
.\"				 conflict with the man page tmac.an file.
.\" 03/10/1998	ARTHUR N.	Added the shortcuts for options.
.\" 06/01/2003	ARTHUR N.	Updated for the new release, and made other,
.\"				 minor enhancements.
.\"
.\"
.\" M4 definitions:
include(tmac.asn)dnl
define(m4_concat, $1$2$3$4$5$6$7$8$9)dnl
define(CURRENT_DATE, translit(esyscmd(date), `
', `'))dnl
define(OVL_VERSION, translit(esyscmd(awk '{print $1; exit}' Version), `
', `'))dnl
.\"
.\"
.\"
.ad l
.TH ovl 8 "CURRENT_DATE" "Arthur Naseef" "Linux System Admin"
.\"
.\" SECTION BREAK
.\"
.SH NAME
ovl.o -
overlay pseudo filesystem.
.\"
.\" SECTION BREAK
.\"
.SH VERSION
This manual page covers version OVL_VERSION of the overlay filesystem.
.\"
.\" SECTION BREAK
.\"
.SH SYNOPSIS
modprobe ovl.o
.br
mount -t ovl \fIovl0\fR /mnt -o root=\fIrootdir1\fR,storage=\fIrootdir2\fR
.\"
.\" SECTION BREAK
.\"
.SH DESCRIPTION
The overlay filesystem is a pseudo filesystem that allows any filesystem,
including one that is read-only, to be modified by saving the changes in
another filesystem.
.P
There is much documentation on the issues of using the overlay filesystem
that is included with the source distribution.
This manual page is intended only to clarify how to use the filesystem.
.\"
.\" SECTION BREAK
.\"
.SH USAGE
In order to use the overlay filesystem,
determine the root directories of the base and storage filesystems,
and any special mount options to use.
Then, the mount command may be used to mount the filesystem.
As shown in the SYNOPSIS above,
use the mount command to mount the filesystem.
The device name is not important, and can be chosen at your leisure,
because the filesystem does not use a device.
.P
For example, if the root of the base filesystem is \fB/cdrom\fR
the root of the storage filesystem is \fB/cdrom_ovl\fR,
the storage file is \fB/.cdrom_stg\fR,
and the mount point is \fB/mnt\fR,
then the following mount command can be used:
.P
.in +4n
.hy 0
.HI /sbin/mount
\-t ovl cdrom_ovl \fB/mnt\fR \-o root=\fB/cdrom\fR,storage=\fB/cdrom_ovl\fR,\
stg_file=\fB/.cdrom_stg\fR
.HE
.hy 1
.in
.\"
.\" SUB-SECTION BREAK
.\"
.SS MOUNT OPTIONS
Many mount options are available
with the ovlfs.
Here are the general-purpose
mount options:
.P
.in +1i
base_root,
storage_root,
stg_method,
nostorage,
maxmem,
xmnt,
noxmnt,
updmntonly,
and noupdmntonly.
.in
.P
Options which affect
the storage of inode mappings,
if compiled
into the driver,
are listed below:
.P
.in +1i
storemaps,
nostoremaps,
basemap,
nobasemap,
stgmap,
and nostgmap
.in
.P
Last, the following options are available for the magic directories,
if compiled into the driver:
.P
.in +1i
magic,
nomagic,
hidemagic,
showmagic,
basemagic,
nobasemagic,
ovlmagic,
noovlmagic,
smagic,
and bmagic
.in
.P
Note that there are shortcuts for these options;
see the subsection, \fBShortcuts\fR, for more information.
Each option is described
in more detail here:
.TP
root=\fIdir\fR
Set the root of the base filesystem to the named directory;
this is the only way to specify this information.
The string \fBbase_root=\fR may be used in place of root=.
.TP
storage=\fIdir\fR
Set the root of the storage filesystem to the named directory;
this is the only way to specify this information.
The string \fBstorage_root=\fR may be used in place of storage=.
.TP
stg_file=\fIpath\fR
Set the path to the storage file to the named path.
All inode information and inode mapping information is
saved between mounts in the storage file.
.TP
nostorage
Turn off the use of storage.
This option does not work yet.
.TP
maxmem
Turn on checking of the amount of available memory before
allocating more memory for inode storage for the overlay filesystem.
From testing, it seems that this option is better left off.
.TP
xmnt
Turn on the crossing of mount points in the base
and storage filesystems.
See the file, Options, distributed with
the source tree for more information.
.TP
noxmnt
Turn off the crossing of mount points in the base
and storage filesystems.
See the file, Options, distributed with
the source tree for more information.
.TP
storemaps
Store all inode mapping information in the storage file.
.TP
nostoremaps
Do not store any inode mapping information in the storage file.
.TP
basemap
Store the inode mapping information for the base filesystem
in the storage file.
.TP
nobasemap
do not store the inode mapping information for the base filesystem
in the storage file.
.TP
stgmap
Store the inode mapping information for the storage filesystem
in the storage file.
.TP
nostgmap
Do not store the inode mapping information for the storage filesystem
in the storage file.
.TP
hidemagic
Do not display the magic directories in directory listings.
In other words, with this option,
the readdir() function will not include the magic directories.
.TP
showmagic
Display the magic directories in directory listings.
This is the opposite of the hidemagic option.
.TP
magic
Turn on the default magic directory options.
At the time of this writing, the default options include
basemagic, ovlmagic, and hidemagic.
.TP
basemagic
Turn on the magic directory for the base filesystem.
.TP
nobasemagic
Turn off the magic directory for the base filesystem.
.TP
ovlmagic
Turn on the magic directory for the storage filesystem.
.TP
noovlmagic
Turn off the magic directory for the storage filesystem.
.TP
smagic=\fIdir\fR
Set the name of the magic directory for the storage filesystem
to the specified name.
Do not include any slashes in this name.
The option, \fBmagic=\fIdir\fR, is a synonym for this option.
.TP
bmagic=\fIdir\fR
Set the name of the magic directory for the base filesystem
to the specified name.
Do not include any slashes in this name.
The option, \fBmagicr=\fIdir\fR, is a synonym for this option.
.\"
.\" SUB-SECTION BREAK
.\"
.SS Shortcuts
All of the supported mount options have shortcuts to
help reduce the length of the mount options on the command line.
.P
.de li
.LI
\\$1
.Tf 24n
\\$2
..
.BL
.li nostorage nost
.li maxmem mm
.li xmnt xm
.li noxmnt noxm
.li updmntonly um
.li noupdmntonly noum
.li base_root=<dir> br=<dir>
.li storage_root=<dir> sr=<dir>
.li stg_file=<path> sf=<path>
.li stg_method=<path> method=<path>
.li storemaps ma
.li nostoremaps noma
.li basemap bm
.li nobasemap nobm
.li stgmap sm
.li nostgmap nosm
.li hidemagic hm
.li magic mg
.li nomagic nomg
.li basemagic mb
.li nobasemagic nomb
.li ovlmagic mo
.li noovlmagic nomo
.li showmagic ms
.li smagic=<name> sn=<name>
.li bmagic=<name> bn=<name>
.LE
.\"
.\" SECTION BREAK
.\"
.SH WARNINGS
The overlay filesystem in many ways emulates the VFS.
There are many possible ways that the use of the overlay
filesystem may cause problems.
A few of the most important issues are mentioned here.
Please see the rest of the documentation included with the source
for more information.
.P
Note that in spite of all of these warnings, and all others
documented for the overlay filesystem,
the use of the overlay filesystem should not cause any corruption
to any of the filesystems involved.
However, for your own safety, it is a good idea to backup ALL
filesystems on the host before using the overlay filesystem.
.P
.BL
.LI
Do not attempt to use a storage root that is a subdirectory of the
base root directory, or vice versa.
.P
.LI
All inodes that are not regular files or directories,
such as device files, pipes, and sockets,
may be directly accessed for reading and writing
through the overlay filesystem.
.P
.LI
The access and change times of the inodes in the base filesystem
may be affected by the use of the overlay filesystem,
although the modification times should not be affected -
except for files that are not directories or regular files.
.P
.LI
Any change to a file in the base filesystem causes the entire
file to be copied into the storage filesystem, even if only
one byte of the file is modified, and regardless of the size
of the file.
.P
.LI
Since names are only pointers to inodes, files and directories that
need to be created in the storage filesystem to store changes from
the base filesystem, may take a name other than the one being
accessed in the base filesystem.
This should only happen with files that have more than one hard link.
.P
.LI
Removing files from the overlay filesystem does not remove their inodes.
The only way to cleanup old inodes in the overlay filesystem is to remove
the storage file, or use a new storage file, for the filesystem.
.LE
.\"
.\" SECTION BREAK
.\"
.SH SEE ALSO
mount(8), ovlfsmount(8), ovlfstab(5)
.\"
.\" SECTION BREAK
.\"
.SH COPYRIGHT
This manual page and associated source code is licensed under the
GNU General Public License Version 2.
Of course, all source derived from the linux kernel source is copyrighted
by the original author of that source.
.sp 1
COPYRIGHT (C) 1998-2003 By Arthur Naseef
